### ISSUE TITLE
Parallel Requests

### ISSUE THREAD

**Issue by luposky** [nats-io/nats.net.v1#231]
I don't understand if this is a bug, or I am just using the client library in a wrong way ... 
The scenario is that I have a service task (connection1 in this example) that reply to request from various source in parallel.
When I have just a normal "for" loop for connection2, all went fine, but If I use Parellel.For strange exception happens and the responses are wrong. I eventually get the message ```(dataIn.QueryIn != dataOut.QueryOut)``` and this is absolutely wrong in my opinion.

The exception i got is ```NATS.Client.NATSException: Unable to deserialize message.``` (from EncodedConn.cs, line 243). This is strange cause the class is perfectly serializable and the error show up only when parallelism is "high" (if 10 parallel requests can be considered "high").

this is the code:

```
using System;
using System.Threading;
using System.Threading.Tasks;
using NATS.Client;


namespace NATSPublisher
{
    class Program
    {
        static void Main(string[] args)
        {
            var random = new Random();
            var connection1 = new ConnectionFactory().CreateEncodedConnection("nats://inner-nats01.test.com:4242");

            connection1.SubscribeAsync("test1", (sender, eventArgs) =>
            {
                // take some time to respond, use a separate task
                Task.Run(() =>
                {
                    try
                    {
                        if (eventArgs.ReceivedObject is LongDataClassResponse dataIn)
                        {
                            Thread.Sleep(1000 + random.Next(100, 1000));
                            connection1.Publish(eventArgs.Reply,
                                new LongDataClassResponse() {QueryIn = dataIn.QueryIn, QueryOut = dataIn.QueryIn});
                        }
                        else
                        {
                            Console.WriteLine("=======================< (null dataIn)");
                        }
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine("=======================< {0}: {1}", ex.GetType(), ex.Message);
                    }
                });
            });

            var connection2 = new ConnectionFactory().CreateEncodedConnection("nats://inner-nats01.test.com:4242");
            Parallel.For(0, 10, i =>
            {
                try
                {
                    var dataIn = new LongDataClassResponse() {QueryIn = i.ToString()};

                    if (connection2.Request("test1", dataIn, 9000) is LongDataClassResponse dataOut)
                    {
                        Console.WriteLine(i);

                        if (dataOut.QueryIn != dataOut.QueryOut) Console.WriteLine("=======================> (dataOut.QueryIn != dataOut.QueryOut)");

                        if (dataIn.QueryIn != dataOut.QueryOut) Console.WriteLine("=======================> (dataIn.QueryIn != dataOut.QueryOut)");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine("=======================> {0}: {1}", ex.GetType(), ex.Message);
                }
            });

            Console.WriteLine("press any key");
            Console.ReadKey();

            connection1.Close();
            connection2.Close();
        }

        [Serializable]
        public class LongDataClassResponse
        {
            public string QueryIn { get; set; }
            public string QueryOut { get; set; }
        }


    }
}

```

any help very appreciated .... 

Best regards,
Luca

**Comment by ColinSullivan1** [comment]
@luposky , Thanks for using NATS, and creating this issue!  I'll take a look.  It sounds like there could be a race condition in the encoded conn someplace.

**Comment by ColinSullivan1** [comment]
@luposky , I did find some threadsafety issues in encoded connections, and will fix them.  We made some changes around requests awhile back and I missed how it could effect the encoded conn.  Truthfully, I've been thinking of deprecating the encoded connection.  Thanks again for raising this issue.   In the meantime, here is code that does the same thing using a standard NATS connections that you can use, so you won't need to wait for the next release.

```csharp
using System;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using NATS.Client;


namespace ParallelEncoder
{
    class CustomEncoder
    {
        private object sStreamLock = new object();
        private MemoryStream sStream = new MemoryStream();
   
        private MemoryStream dStream = new MemoryStream();
        private object dStreamLock = new object();

        private IFormatter f = new BinaryFormatter();

        // byte array in the connection, we are still threadsafe.
        internal byte[] ObjSerializer(Object obj)
        {
            if (obj == null)
                return null;

            lock (sStreamLock)
            {
                sStream.Position = 0;
                f.Serialize(sStream, obj);

                long len = sStream.Position;
                var sBytes = new byte[len];

                Array.Copy(sStream.GetBuffer(), sBytes, len);
                return sBytes;
            }
        }

        internal object ObjDeserializer(byte[] data)
        {
            if (data == null)
                return null;

            lock (dStreamLock)
            {
                dStream.Position = 0;
                dStream.Write(data, 0, data.Length);
                dStream.Position = 0;

                return f.Deserialize(dStream);
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var ce = new CustomEncoder();

            var random = new Random();
            var connection1 = new ConnectionFactory().CreateConnection("nats://127.0.0.1:4222");

            connection1.SubscribeAsync("test1", (sender, eventArgs) =>
            {
                // take some time to respond, use a separate task
                Task.Run(() =>
                {
                    try
                    {
                        var ReceivedObject = ce.ObjDeserializer(eventArgs.Message.Data);
                        if (ReceivedObject is LongDataClassResponse dataIn)
                        {
                            Thread.Sleep(1000 + random.Next(100, 1000));
                            connection1.Publish(eventArgs.Message.Reply,
                                ce.ObjSerializer(new LongDataClassResponse() { QueryIn = dataIn.QueryIn, QueryOut = dataIn.QueryIn }));
                        }
                        else
                        {
                            Console.WriteLine("=======================< (null dataIn)");
                        }
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine("=======================< {0}: {1}", ex.GetType(), ex.Message);
                    }
                });
            });

            var connection2 = new ConnectionFactory().CreateConnection("nats://127.0.0.1:4222");
            Parallel.For(0, 10, i =>
            {
                try
                {
                    var dataIn = new LongDataClassResponse() { QueryIn = i.ToString() };

                    var retObj = ce.ObjDeserializer(connection2.Request("test1", ce.ObjSerializer(dataIn), 9000).Data);
                    if (retObj is LongDataClassResponse)
                    {
                        var dataOut = (LongDataClassResponse)retObj;

                        if (dataOut.QueryIn != dataOut.QueryOut)
                            Console.WriteLine("=======================> (dataOut.QueryIn ({0}) != dataOut.QueryOut ({1}))", dataOut.QueryIn, dataOut.QueryOut);

                        if (dataIn.QueryIn != dataOut.QueryOut)
                            Console.WriteLine("=======================> (dataIn.QueryIn  ({0}) != dataOut.QueryOut ({1}))", dataIn.QueryIn, dataIn.QueryOut);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine("=======================> {0}: {1}", ex.GetType(), ex.Message);
                }
            });

            Console.WriteLine("press any key");
            Console.ReadKey();

            connection1.Close();
            connection2.Close();
        }

        [Serializable]
        public class LongDataClassResponse
        {
            public string QueryIn { get; set; }
            public string QueryOut { get; set; }
        }
    }
}
```

**Comment by luposky** [comment]
thank you @ColinSullivan1 very appreciated !
